devdoc/Metrics.rst

```diff
@@ -0,0 +1,282 @@
+measurements to the collection system. The *Metrics Reporter* provides the support to complete
+these functions.
+
+The sink encapsulates all of the collection system requirements providing a pluggable architecture that
+isolates components from these differences. The framework supports multiple sinks concurrently,
+each operating independently.
+
+Instrumented components are not aware of the sink or sinks in use. Sinks can be changed without
+requiring changes to a component. Therefore, components are independent of the collection system(s) in use.
+
+
+Metrics Reporter
+================
+
+The metrics reporter class provides all of the common functions to bind together the component, the metrics it creates, and the sinks to which measurements are reported. It is responsible for the following:
+
+* Initialization of the framework
+* Managing the metrics created by the component
+* Handling collection and reporting as directed by configured sinks
+
+
+Metrics Implementations
+=======================
+
+The sections that follow discuss metric implementations.
+
+Gauge Metric
+------------
+A gauge metric is a continuously updated value representing the current state of an interesting value in the component. For example, the amount of memory used in an internal buffer, or the number of requests waiting on a queue. A gauge metric may increase or decrease in value as needed. Reading the value of a gauge is a stateless operation in that there are no dependencies on the previous reading. The value returned shall always be the current state.
+
+Once created, the component shall update the gauge anytime the state of what is measured is updated. The metric shall provide methods to increase and decrease the value. The sink reads the value during collection an reporting.
+
+
+*************
+Configuration
+*************
+This section discusses configuration. Since Helm charts are capable of combining configuration data at a global level into a component's specific configuration, a single config source is all all that is required. The configuration takes the form as shown below. Note that as the design progresses it is expected that there will be additions.
+
+::
+
+  component:
+    metrics:
+      sinks:
+        - name: <sink name>
+          type: <sink_type>
+          settings:
+            sink_setting1: sink_setting_value1
+            sink_setting2: sink_setting_value2
+          metrics:
+            - name: <metric_name>
+
+Where (based on being a child of the current *component*):
+
+metrics
+    Metrics configuration for the component
+
+metrics.sinks
+    List of sinks defined for the component (may have been combined with global config)
+
+metrics.sinks.name
+    A name for the sink. Note this may not be needed, but can provide a way to combine global and component config based on value
+
+metrics.sinks.type
+    The type for the sink. The type defines how to find the sink in order to load it. The value is used as follows:
```

Comment:
There should be a simple standard mapping from the type to the name of the file - which should be deployed in our standard location for shared objects.  e.g. prometheus -> libprometheousmetric.so.  I don't think we need to offer options- unless they solve a design issue.

Reply:
We can certainly simplify the mapping (using #2 probably). However, @rpastrana originally felt it was important to be able to fully specify the name of the sink in case the sink support was contained in a lib that also provided other support. It also provided a way to quickly test a new sink w/o having to save and restore the old one. Before removing the ability to fully specify the name I just wanted to identify the reasons it was added.

Reply:
I'm not really convinced by either example.